### PR TITLE
Add rom2: support (Fix Chinese Bios)

### PR DESCRIPTION
**Pull request by kozarovv**
Add support for loading rom2: . 
This is required by Chinese SCPH-50009 bios, and by at least 3 games. 
This was tested by myself with non Chinese 2.00 rom from my PS2. without .rom2, to make sure it won't break stuff.

And finally i was able to find user with SPCH-50009 rom on one of ps2 related discords, that was able to test it for me. Seems to work fine. Additional feedback here: https://github.com/PCSX2/pcsx2/issues/3141

Basing on redump data, i compiled list of games that can possibly get profit from this pull request. 
Ticked entries mean that PR help/fix game, unticked that game wasn't tested yet (possibly not all games need it).  

- [x] Ape Escape 2 - 捉猴啦 2 (SCCS 40001) - **Fixed by PR**
- [ ] Ape Escape: Pumped & Primed - 比波猴百斗派对 (SCCS 40016)
- [ ] Arc the Lad: Seirei no Tasogare - 亚克传承　～精灵之黄昏～ (SCCS 40007)
- [ ] Armored Core 2: Another Age - 机战佣兵2 (SCCS 40011)
- [ ] Devil May Cry 2 (Disc 1) - 恶魔猎人2 (SCCS 40002)
- [ ] Devil May Cry 2 (Disc 2) - 恶魔猎人2 (SCCS 40003)
- [ ] Dragon Ball Z 2 - 龙珠二世 (SCCS 40009)
- [ ] EyeToy: Play - 爱淘儿 (SCCS 40017)
- [ ] EyeToy: Saru EyeToy: Oosawagi! Ukkiuki Game Tenkomori!! - 比波猴@爱淘儿 (SCCS 40018)
- [ ] Formula One 04 - F1 04 (SCCS 40019)
- [ ] Ico ICO - 古堡迷踪 (SCCS 40005)
- [ ] Shin Sangoku Musou 2 - 真·三国無双2 (SCCS 40006)
- [ ] Super Puzzle Bobble 2 - 超级泡泡龙 2 (SCCS 40010)
- [ ] Viorate no Atelier: Gramnad no Renkinjutsushi 2 - 薇欧蕾特的工房 ～古拉姆那特的炼金术士2～ (SCCS 40015)
- [ ] World Soccer Winning Eleven 7: International - 实况足球 胜利十一人7 国际版 (SCCS 40014)
- [ ] World Soccer Winning Eleven 8: Asia Championship - 实况足球 胜利十一人 8 亚洲冠军联赛 (SCCS 40022)
- [X] XIGO: Zuihou de Touzi XIGO: 最后的骰子 - XI GO: Zuì​hòu​ de Tóu​zi​ (SCCS 40004) - **Fixed by PR**
- [X] Gran Turismo 4 - 跑车浪漫旅 4 (SCCS 60002) - **Fixed by PR**

### CODE REVIEW

File: pcsx2/IopMem.cpp [hunk 1]
```diff
@@ -96,6 +96,10 @@ void iopMemoryReserve::Reset()
 		psxMemWLUT[i + 0x2000 + 0x1e00] = (uptr)&eeMem->ROM1[i << 16];
 	}
 
+	for (int i = 0; i < 0x0008; i++) {
+        psxMemWLUT[i + 0x2000 + 0x1e40] = (uptr)&eeMem->ROM2[i << 16];
+    }
```
Comment (lightningterror) [inline pcsx2/IopMem.cpp:101]: Would be nice to fix the formatting throughout the pr. Files use tabs for indentation.

Comment (kozarovv) [reply pcsx2/IopMem.cpp]: Should be ok now for all files.

**Comment by gtgamer468** [comment]
I had a friend test his GT4 Chinese Press Copy with this PR as it never worked in the master for him. This PR fixes the booting issues he'd get. The build he used was this one: [https://github.com/PCSX2/pcsx2/issues/3141#issuecomment-639382625]

**Comment by Qwest8K** [comment]
Sorry if I writing in wrong place.

In "Q3 2020 Progress Report" in the paragraph of adding support for loading rom2, was said that this allowed the launch games of Chinese region. And three of Chinese game work for sure. And other games on could be fixed as well, someone just needs to test/verify them.

So I made express test of all remaining games. Before, I want to note that not a single game can launched through a full load of the Chinese BIOS - only a fast one. Settings was default (except disabled Texture Filtering of Display and Automatic Gamefixes), OpenGL 4x Native, all speedhucks disabled:
1) Ape Escape: Pumped & Primed (SCCS 40016) CRC=D2B89DC8
Ape Escape - Pumped   Primed
Works through every 60 FPS BIOS (USA and Japan with full and fast load) with strong speed drops (GS loads).
2) Arc the Lad: Seirei no Tasogare (SCCS 40007) CRC=8868C43A
Arc the Lad - Seirei no Tasogare
Works only through Chinese BIOS. At full speed. In hardware mode has a lot of graphic artifacts.
3) Armored Core 2: Another Age (SCCS 40011) CRC=C65F2626
Armored Core 2 - Another Age
Works only through Chinese BIOS. With huge speed drops. Loads EE and GS both.
4) Devil May Cry 2 (Disc 1) (SCCS 40002) CRC=F6EFF947
Devil May Cry 2 (Disc 1)
Works only through Chinese BIOS. At full speed.
5) Devil May Cry 2 (Disc 2) (SCCS 40003) CRC=F6EFF947
Devil May Cry 2 (Disc 2)
Works only through Chinese BIOS. At full speed.
6) Dragon Ball Z 2 (SCCS 40009) CRC=C6ACEE82
Dragon Ball Z 2
Works through every 60 FPS BIOS with maximum load of EE.
7) EyeToy: Play (SCCS 40017) CRC=310F5035
EyeToy - Play
Works only through Chinese BIOS. Obviously, requires an EyeToy for more tests.
8) EyeToy: Saru EyeToy: Oosawagi! Ukkiuki Game Tenkomori!! (SCCS 40018) CRC=D196EC2B
EyeToy - Saru EyeToy Oosawagi! Ukkiuki Game Tenkomori!!
Works only through Chinese BIOS. Requires an EyeToy.
9) Formula One 04 (SCCS 40019) CRC=17414B45
Formula One 04
Works only through Chinese BIOS. With huge speed drops (EE loads). Suddenly has English localization.
10) Ico ICO (SCCS 40005) CRC=BF3A46DA
[Ico ICO (emuLog).txt](https://github.com/PCSX2/pcsx2/files/5380183/Ico.ICO.emuLog.txt)
Doesn't start. Completely.
11) Shin Sangoku Musou 2 (SCCS 40006) CRC=25E9141A
Shin Sangoku Musou 2
Works only through Chinese BIOS. With speed drops and graphical artifacts in hardware mode.
12) Super Puzzle Bobble 2 (SCCS 40010) CRC=9F744707
Super Puzzle Bobble 2
Works only through Chinese BIOS. At full speed.
13) Viorate no Atelier: Gramnad no Renkinjutsushi 2 (SCCS 40015) CRC=E9025F99
Viorate no Atelier - Gramnad no Renkinjutsushi 2
Works only through Chinese BIOS. At full speed. Prerendered videos get artifacts at resolutions higher than native.
14) World Soccer Winning Eleven 7: International (SCCS 40014) CRC=3C482020
World Soccer Winning Eleven 7 - International
Works only through Chinese BIOS. At full speed.
15) World Soccer Winning Eleven 8: Asia Championship (SCCS 40022) CRC=F03F7202
[World Soccer Winning Eleven 8 - Asia Championship (emuLog).txt](https://github.com/PCSX2/pcsx2/files/5380188/World.Soccer.Winning.Eleven.8.-.Asia.Championship.emuLog.txt)
Doesn't start. Completely.

**Comment by AKuHAK** [comment]
It appears the issue might not be directly related to PCSX2. After investigating, I discovered that all current dumps of ROM2 were incorrect, specifically being underdumps. The standard ROM2 size found in most internet archives is 512KB, which is not accurate. The SCPH-50009 model actually uses a larger DVDROM chip, which is 8MB in size. Consequently, the correct ROM2 size should be 4MB.

Upon further inspection, the actual data occupies approximately 1.5MB. I've cropped ROM2 to 2Mb and uploaded the corrected version to archive.org. You can find it by searching for "SCPH-50009 BIOS" from @israpps (El_isra).

Given that more than 1MB of data was previously missing, it would be beneficial to retest any problematic games using the full BIOS data.

**Comment by refractionpcsx2** [comment]
Mention piracy on our github again and I will remove you from it.

**Comment by AKuHAK** [comment]
My apologizes. Corrected file data:
size 2Mb
md5 c48c4bacb5baf48226e5379a531247fa

**Comment by Qwest8K** [comment]
@AKuHAK 
Both problematic games was already retested and they successfully run on uncorrected BIOS version: [World Soccer Winning Eleven 8](https://forums.pcsx2.net/Thread-World-Soccer-Winning-Eleven-8-Asia-Championship-SCCS-40022-C?amp%3Baction=lastpost) and [ICO](https://forums.pcsx2.net/Thread-ICO-SCCS-40005-C?amp%3Baction=lastpost).

But I rechecked this two and other Chinese games with the full BIOS data, and got nothing new - all games are playable, runs only with Chinese BIOS and only in Fast Boot. In Full Boot mode still appearing a message "Please insert PS or PS2 format disc".

**Comment by Goatman13** [comment]
Old SCPH-50009 dumps were missing great chunk of GB18030 font file, which file itself should be 1,963,188 bytes in size.  Only two games are known to really use this font: SCCS-40001 Ape Escape 2, and SCCS-40015 Viorate no Atelier: Gramnad no Renkinjutsushi 2. Other games just test if rom2 exist but never really use it. That don't include Eye Toy games as at testing time they were unplayable.

I'm leaving this info here just in case if there is ever bug report for those games. With 2MB of fonts who knows which parts were missing in old dump, maybe later in-game some characters will be broken on incomplete bios.

**Comment by AKuHAK** [comment]
@Goatman13 thanks for detailed report. Currently PCSX2 has no size limits for any BIOS module, it will try to load them at hardcoded addresses until there is free memory. This is not how real PS2 is doing this, address is hardcoded for rom0 only, but as those addresses are always the same its safe to assume that pcsx2 is correctly doing this.
https://github.com/PCSX2/pcsx2/blob/cd3e11bff77bb5a7f44c5a39cb534b670c09abd7/pcsx2/ps2/BiosTools.cpp#L376

**Comment by AKuHAK** [comment]
Or do you mean this piece of code? I agree that it looks suspicious, but I am not sure what actually it is doing.
https://github.com/PCSX2/pcsx2/blob/master/pcsx2/IopMem.cpp#L70-L85

**Comment by AKuHAK** [comment]
@Goatman13 thanks, the IOP memory rom2 region was really limited to 512KB. Since games mostly use IOP memory for loading ROM data, only after https://github.com/PCSX2/pcsx2/pull/11572 is merged will there be any noticeable differences compared to the previous state.

**Pull request merged**